### ISSUE TITLE
[Bug Report] enable_cache leads to hang if run nanogpt model training

### ISSUE THREAD

**Issue by dmakoviichuk-tt** [tenstorrent/tt-metal#13822]
**Describe the bug**
If run nanogpt training https://github.com/tenstorrent/TT-Tron/blob/main/sources/examples/nano_gpt/main.cpp
It hangs after 2 training step if call `device->enable_program_cache();`
Currently `device->enable_program_cache();` is called only in eval mode and it gives significant speedup. But we cannot use it during training. For other and smaller models like mnist it works fine.
To test it: please move this code out of eval check: https://github.com/tenstorrent/TT-Tron/blob/39c560b6e0e92f71f81669dfef4ad3be35a6275a/sources/examples/nano_gpt/main.cpp#L257C9-L257C40

**To Reproduce**
1. Run nanogpt training with cache enabled in the https://github.com/tenstorrent/tt-tron repo
2. Wait for the 2 training steps and it will hang just after the second step. Seems like it is deadlock inside of the CommanQueue.
3. If you reduce number of blocks:     `uint32_t num_blocks = 6;` in the https://github.com/tenstorrent/TT-Tron/blob/39c560b6e0e92f71f81669dfef4ad3be35a6275a/sources/examples/nano_gpt/main.cpp#L37
It will work or might hang later.

**Expected behavior**
caching should work.

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit)

**Additional context**
enable caching gives significant performance boost so it is a P0 for us. Current training time is near 4 hours and it is hard to iterate.
Similar pytorch code on 3090 takes 3-4 minutes to converge.
If you have any issues with tt-tron code feel free to reach me or @rfurko-tt.

**Comment by dmakoviichuk-tt** [comment]
Hi @TT-BrianLiu  unfortunately we found that it is embedding_bw op.
For some reason if reuse it from cache it hangs the app. 
Could you help us please?
